stumpy/stump_tiles.py

```diff
@@ -0,0 +1,780 @@
+        `False`. Default is `True`.
+
+    k : int
+        The number of top `k` smallest distances used to construct the matrix profile.
+        Note that this will increase the total computational time and memory usage
+        when k > 1.
+
+    Returns
+    -------
+    None
+
+    Notes
+    -----
+    `DOI: 10.1007/s10115-017-1138-x \
+    <https://www.cs.ucr.edu/~eamonn/ten_quadrillion.pdf>`__
+
+    See Section 4.5
+
+    The above reference outlines a general approach for traversing the distance
+    matrix in a diagonal fashion rather than in a row-wise fashion.
+
+    `DOI: 10.1145/3357223.3362721 \
+    <https://www.cs.ucr.edu/~eamonn/public/GPU_Matrix_profile_VLDB_30DraftOnly.pdf>`__
+
+    See Section 3.1 and Section 3.3
+
+    The above reference outlines the use of the Pearson correlation via Welford's
+    centered sum-of-products along each diagonal of the distance matrix in place of the
+    sliding window dot product found in the original STOMP method.
+    """
+    m_inverse = 1.0 / m
+    constant = (m - 1) * m_inverse * m_inverse  # (m - 1)/(m * m)
+    uint64_m = np.uint64(m)
+
+    # for storing diagonal iteration ranges
+    iter_ranges = np.zeros((config.STUMPY_N_DIAGONALS, 3), dtype=np.int64)
+    # for storing covariance values while iterating in chunks
+    covariances = np.zeros(config.STUMPY_N_DIAGONALS, dtype=np.float64)
+    # for storing range of diagonals to traverse in chunks
+    chunk_diags_range = np.arange(config.STUMPY_N_DIAGONALS)
+
+    for tile_idx in range(tiles_ranges[thread_idx, 0], tiles_ranges[thread_idx, 1]):
+        (
+            tile_i,  # i location of current tile in distance profile
+            tile_j,  # j location of current tile in distance profile
+            tile_height,  # height of current tile
+            tile_width,  # width of current tile
+            tile_start_diag,  # index of first diagonal to traverse on current tile
+            tile_stop_diag,  # index of last diagonal to traverse on current tile + 1
+            _,
+        ) = tiles[tile_idx]
+
+        # chunk_start_idx is the starting index of the chunk we are dealing with
+        for chunk_start_idx in range(
+            tile_start_diag,
+            tile_stop_diag,
+            config.STUMPY_N_DIAGONALS,
+        ):
+            # number of diagonals to traverse together
+            chunk_size = min(
+                config.STUMPY_N_DIAGONALS,
+                tile_stop_diag - chunk_start_idx,
+            )
+            # range of diags to traverse in chunk
+            chunk_diags = chunk_diags_range[:chunk_size]
+            # range of diags to traverse in chunk, with respect to entire distance
+            # matrix
+            tile_diags = chunk_start_idx + chunk_diags
+            # store the diagonal indices we are dealing with, with respect to entire
+            # distance matrix
+            iter_ranges[chunk_diags, 0] = tile_diags - tile_i + tile_j
```

Comment:
This is related to my last comment. 

I think the intermediate variable  `chunk_diags` can be avoided here. Previously, it was suggested when we try to vectorize the computation of  `iter_ranges.` 

Now that I think about it, **it seems that it can force the reader to go back and check `chunk_diags`**.  Although `chunk_diags = chunk_diags_range[:chunk_size]` and its use cases are close in the code, it may create a little bit of difficulty for the reader.

```
tile_diags = chunk_start_idx + chunk_diags_range[:chunk_size]
iter_ranges[:chunksize, 0] = tile_diags + ...
```

(btw, the downside is that we need to repeat `:chunk_sisze` for the next few lines, but the good thing about this is  that it can show that we are dealing with the first `chunk_size` elements (and not just any array of indices)
Again, this is personal opinion. 

@alvii147 @seanlaw 
What do you think?

Reply:
@NimaSarajpoor good point, I'm honestly split here, it would be nice to get rid of an intermediate variable here, but then again it might make things very slightly less readable. @seanlaw can you take a look at this and let me know what you think? I'm okay with either way.

Reply:
@alvii147 I'll take a look altogether. Maybe leave it alone for now and I'll try to review this comment (or remind me if I miss it) when I look at everything in one go. How does that sound?